### PR TITLE
deploymentapi: create refactor and Enterprise Search support

### PR DESCRIPTION
**Pull request by karencfv**
## Description
This patch refactors the resources create APIs in order to avoid
making multiple unnecessary API calls when creating a new deployment.

In preparation of switching over to the new deployment template APIs,
these changes also include using the `deployment` format instead of
`cluster` as this will now be the default.

Additionally deployment create now supports Enterprise Search

In a follow-up PR, these APIs will call the new deployment templates API 

## Related Issues
Part of https://github.com/elastic/ecctl/issues/334

## Motivation and Context
Necessary for ecctl GA

## How Has This Been Tested?
Manually by faking the SDK in ecctl and creating several deployments, and with unit tests.

## Types of Changes

- [x] Refactoring (improves code quality but has no user-facing effect)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Readiness Checklist

- [x] My code follows the code style of this project
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

### CODE REVIEW

File: pkg/api/deploymentapi/depresourceapi/enterprise_search.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+		return nil, err
+	}
+
+	// When either not set, we obtain the values from the running deployment.
+	// Overriding either or both is done at the end of the if.
+	if err := getTemplateAndRefID(&params); err != nil {
+		return nil, err
+	}
+
+	if params.DeploymentTemplateInfo.DeploymentTemplate.Resources.EnterpriseSearch == nil {
```
Comment (marclop) [inline pkg/api/deploymentapi/depresourceapi/enterprise_search.go:47]: I'd also move this check to `len() == 0`.

File: pkg/api/deploymentapi/depresourceapi/new_deployment_test.go [hunk 2]
```diff
@@ -463,16 +529,107 @@ func TestNew(t *testing.T) {
...
+					RefID:     "main-enterprise_search",
+					Size:      1024,
+					ZoneCount: 1,
+				},
+				DeploymentTemplateID:   "default",
+				EnterpriseSearchEnable: true,
+				API: api.NewMock(
+					mock.New200Response(mock.NewStructBody(enterpriseSearchKibanaTemplateResponse)),
+					mock.New200Response(mock.NewStructBody(enterpriseSearchKibanaTemplateResponse)),
+					mock.New200Response(mock.NewStructBody(enterpriseSearchKibanaTemplateResponse)),
```
Comment (marclop) [inline pkg/api/deploymentapi/depresourceapi/new_deployment_test.go:557]: Only 1 API call is required now
```suggestion
					mock.New200Response(mock.NewStructBody(enterpriseSearchKibanaTemplateResponse)),
```

File: pkg/api/deploymentapi/depresourceapi/elasticsearch.go [hunk 3]
```diff
@@ -116,25 +115,16 @@ func (params *NewElasticsearchParams) Validate() error {
...
+	err := params.Validate()
 	if err != nil {
 		return nil, err
 	}
 
 	var payload = newElasticsearchPayload(params)
 	payload.Plan.ClusterTopology, err = BuildElasticsearchTopology(
 		BuildElasticsearchTopologyParams{
-			ClusterTopology: res.ClusterTemplate.Plan.ClusterTopology,
+			ClusterTopology: params.DeploymentTemplateInfo.DeploymentTemplate.Resources.Elasticsearch[0].Plan.ClusterTopology,
```
[the remark this answers is not in the document]
Comment (karencfv) [reply pkg/api/deploymentapi/depresourceapi/elasticsearch.go:132]: all deployment templates have ES but I'll add the check anyway :)

File: pkg/api/deploymentapi/depresourceapi/elasticsearch.go [hunk 4]
```diff
@@ -116,25 +115,16 @@ func (params *NewElasticsearchParams) Validate() error {
...
 
-	res, err := configurationtemplateapi.GetTemplate(configurationtemplateapi.GetTemplateParams{
-		API:                params.API,
-		ID:                 params.TemplateID,
-		Region:             params.Region,
-		Format:             "cluster",
-		ShowInstanceConfig: true,
-	})
+	err := params.Validate()
 	if err != nil {
```
[the remark this answers is not in the document]
Comment (karencfv) [reply pkg/api/deploymentapi/depresourceapi/elasticsearch.go:120]: If I do it that way I get an `undefined: err` but if I declare `var err error` beforehand the linter complains 😄

**Pull request merged**